src/lexer/strings.js

```diff
@@ -126,25 +129,100 @@ module.exports = {
   },
 
   // check if its a DOC end sequence
-  isDOC_MATCH: function() {
+  isDOC_MATCH: function(offset, consumeLeadingSpaces) {
     // @fixme : check if out of text limits
+
+    // consumeLeadingSpaces is false happen DOC prematch END HEREDOC stage.
+
+    // Ensure current state is really after a new line break, not after a such as ${variables}
+    const prev_ch = this._input.substring(offset - 2, offset - 1);
+    if (prev_ch !== "\n" && prev_ch !== "\r") {
+      return false;
+    }
+
+    // skip leading spaces or tabs
+    let indentation_uses_spaces = false;
+    let indentation_uses_tabs = false;
+    // reset heredoc_label structure
+    let indentation = 0;
+    let leading_ch = this._input.substring(offset - 1, offset);
+
+    while (/[\t ]/.test(leading_ch)) {
```

Comment:
please, no regex as they perform about 10 times slower than simple char checks `while ( leading_ch == '\t' || leading_ch == ' ') { `

Reply:
Ok.. just removed this regex in loop